Document recovering from invalid merges on composer.lock and composer.json (#11154)
composer/composer: doc/articles/resolving-merge-conflicts.md

To do this, run the following commands:

```shell
composer validate

> composer install [--dry-run]

Reviewer: Please undo this, we use `php composer.phar` as way to invoke composer throughout the docs for consistency. Please also use that style below in the new examples.

to [semantic versioning](https://semver.org/) when using
[next significant release operators](versions.md#next-significant-release-operators) should make sure
that merging branches does not break anything by accidentally updating a dependency.

# Recovering from incorrectly resolved merge conflicts

If the above steps aren't followed and text based merges have been done anyway, your Composer project might be in a state where unexpected behaviour is observed because the `composer.lock` file is not (fully) in sync with the `composer.json` file. There are two things that can happen here:

> 1. There are packages in the `require` or `require-dev` section of the `composer.json` file that are not in the lock file and as a result never installed

Reviewer: Is that true? I'd think with latest composer this state would fail an install from lock.

Author: On the latest version of composer:
```shell
$ composer --version
Composer version 2.4.4 2022-10-27 14:39:29
```

With the following composer.json file:
```json
{
    "name": "prinsfrank/test",
    "description": "non-locked-but-required-package-test",
    "require": {
    },
    "license": ["MIT"]
}
```

And this lock file: 
```json
{
    "_readme": [
        "This file locks the dependencies of your project to a known state",
        "Read more about it at https://getcomposer.org/doc/01-basic-usage.md#installing-dependencies",
        "This file is @generated automatically"
    ],
    "content-hash": "174d0b0103772349fae804b011bcd872",
    "packages": [],
    "packages-dev": [],
    "aliases": [],
    "minimum-stability": "stable",
    "stability-flags": [],
    "prefer-stable": false,
    "prefer-lowest": false,
    "platform": [],
    "platform-dev": [],
    "plugin-api-version": "2.3.0"
}
```

The lock file is valid:
```shell
$ composer validate
./composer.json is valid

$ composer i
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Nothing to install, update or remove
Generating autoload files
```

If we the simulate a text based edit/merge by adding a dependency:
```diff
{
    "name": "prinsfrank/test",
    "description": "non-locked-but-required-package-test",
    "require": {
+        "symfony/console": "^6.1"
    },
    "license": ["MIT"]
}
```

A `composer install` only gives a warning about the outdated lock file hash:
```shell
$ composer i
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. It is recommended that you run `composer update` or `composer update <package name>`.
Nothing to install, update or remove
Generating autoload files
```

Which even disappears when regenerating the lock file hash:
```shell
$ composer update --lock
Loading composer repositories with package information
Updating dependencies
Nothing to modify in lock file
Writing lock file
Installing dependencies from lock file (including require-dev)
Nothing to install, update or remove
Generating autoload files
No installed packages - skipping audit.

$ composer i
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Nothing to install, update or remove
Generating autoload files
```

The only way to detect this state is by running a 'composer validate':
```shell
$ composer validate
./composer.json is valid but your composer.lock has some errors
# Lock file errors
- Required package "symfony/console" is not present in the lock file.
This usually happens when composer files are incorrectly merged or the composer.json file is manually edited.
Read more about correctly resolving merge conflicts https://getcomposer.org/doc/articles/resolving-merge-conflicts.md
and prefer using the "require" command over editing the composer.json file directly https://getcomposer.org/doc/03-cli.md#require
```

We added this behaviour of detecting this state a little over a year ago in #9899. but only to the 'composer validate' command.

Reviewer 2: `composer install` not failing looks like a bug to me. It looks like it does not read the root package requirements properly.

Author: I opened a PR to fix this: https://github.com/composer/composer/pull/11195 If that gets merged or the behaviour has changed elsewhere I'll update this PR. ;)

Reviewer: Please update the PR then whenever you have a sec ;)

Author: @Seldaek What did you have in mind for documenting this change? I have just added a note to the article, as the PR for the extra check was to main so it will be only released in 2.5. Should we backport this to previous versions as well?

Reviewer: The added note look fine to me. I'll not backport this as it is kind of a functionality change/feature rather than fix.

2. There are packages in the `composer.lock` file that are not a direct or indirect dependency of any of the packages required. As a result, a package is installed, even though running `composer why vendor/package` says it is not required.

There are several ways to fix these issues;

## A. Start from scratch

The easiest but most impactful option is to remove your invalid `composer.lock` file and run a `php composer.phar update` or `php composer.phar install` to resolve to a correct state from scratch.

A drawback to this is that previously locked package versions are now updated, as the information about previous package versions has been lost. If all your dependencies follow [semantic versioning](https://semver.org/) and your [version constraints](../articles/versions.md) are using [next significant release operators](versions.md#next-significant-release-operators) this should not be an issue, otherwise you might inadvertently break your application.

## B. Reconstruct from the git history

An option that is probably not very feasible in a lot of situations but that deserves an honorable mention;

> It might be possible to reconstruct the correct package state by going back into the git history and finding the most recent valid composer file, and re-requiring the new dependencies from there.

Reviewer: I think you mean lock here?

```suggestion
It might be possible to reconstruct the correct package state by going back into the git history and finding the most recent valid composer.lock file, and re-requiring the new dependencies from there.
```

to [semantic versioning](https://semver.org/) when using
[next significant release operators](versions.md#next-significant-release-operators) should make sure
that merging branches does not break anything by accidentally updating a dependency.

# Recovering from incorrectly resolved merge conflicts

> If the above steps aren't followed and text based merges have been done anyway, your Composer project might be in a state where unexpected behaviour is observed because the `composer.lock` file is not (fully) in sync with the `composer.json` file. There are two things that can happen here:

Reviewer: ```suggestion
If the above steps aren't followed and text based merges have been done anyway,
your Composer project might be in a state where unexpected behaviour is observed
because the `composer.lock` file is not (fully) in sync with the `composer.json` file.

There are two things that can happen here:
```

1. There are packages in the `require` or `require-dev` section of the `composer.json` file that are not in the lock file and as a result never installed

> > **Note:** Starting from Composer release 2.5, packages that are required but not locked will result in an error when running ```composer install```

Reviewer: ```suggestion
> **Note:** Starting from Composer release 2.5, having packages that are required but not present in `composer.lock` results in an error when running `install`
```
